File: aicq-server/src/api/authRoutes.ts
import { Router, Request, Response } from 'express';
import * as verificationService from '../services/verificationService';
import * as accountService from '../services/accountService';
import { generalLimiter } from '../middleware/rateLimit';
import { loginRateLimit } from '../middleware/auth';

const router = Router();

/**
 * POST /api/v1/auth/send-code
 * Send a verification code to email or phone.
 */
router.post('/auth/send-code', generalLimiter, async (req: Request, res: Response) => {
  try {
    const { target, type, purpose } = req.body;

    if (!target || !type || !purpose) {
      res.status(400).json({ error: 'Missing target, type, or purpose' });
      return;
    }

    if (!['email', 'phone'].includes(type)) {
      res.status(400).json({ error: 'Type must be email or phone' });
      return;
    }

    if (!['register', 'login', 'reset_password'].includes(purpose)) {
      res.status(400).json({ error: 'Purpose must be register, login, or reset_password' });
      return;
    }

    // Basic email/phone validation
    if (type === 'email' && !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(target)) {
      res.status(400).json({ error: '邮箱格式无效' });
      return;
    }

    if (type === 'phone') {
      const phoneStr = String(target);
      if (!/^\d{4,15}$/.test(phoneStr)) {
        res.status(400).json({ error: '手机号格式无效' });
        return;
      }
    }

    const result = await verificationService.sendVerificationCode(target, type, purpose);
    res.json({
      success: true,
      expiresAt: result.expiresAt,
      message: '验证码已发送',
    });
  } catch (err: any) {
    res.status(400).json({ error: err.message });
  }
});

/**
 * POST /api/v1/auth/register
 * Register a new human account.
 */
router.post('/auth/register', generalLimiter, async (req: Request, res: Response) => {
  try {
    const { target, type, code, password, displayName, publicKey } = req.body;

    if (!target || !type || !code || !password) {
      res.status(400).json({ error: 'Missing required fields: target, type, code, password' });
      return;
    }

    if (!publicKey) {
      res.status(400).json({ error: 'Missing publicKey' });
      return;
    }

    if (!['email', 'phone'].includes(type)) {
      res.status(400).json({ error: 'Type must be email or phone' });
      return;
    }

    if (password.length < 6) {
      res.status(400).json({ error: '密码长度至少6位' });
      return;
    }

    const account = await accountService.registerHuman(target, type, code, password, displayName, publicKey);

    // Auto-login after registration
    const session = accountService.createSession(account);

    res.status(201).json({
      success: true,
      account: sanitizeAccount(account),
      session: sanitizeSession(session),
    });
  } catch (err: any) {
    res.status(400).json({ error: err.message });
  }
});

/**
 * POST /api/v1/auth/login
 * Login with email+password or phone+code.
 */
router.post('/auth/login', generalLimiter, loginRateLimit, async (req: Request, res: Response) => {
  try {
    const { target, type, password, code } = req.body;

    if (!target || !type) {
      res.status(400).json({ error: 'Missing target or type' });
      return;
    }

    if (!password && !code) {
      res.status(400).json({ error: 'Missing password or code' });
      return;
    }

    const { account, session } = await accountService.loginHuman(target, type, password, code);

    res.json({
      success: true,
      account: sanitizeAccount(account),
      session: sanitizeSession(session),
    });
  } catch (err: any) {
    res.status(401).json({ error: err.message });
  }
});

/**
 * POST /api/v1/auth/login-agent
 * AI Agent login with public key signature.
 */
router.post('/auth/login-agent', generalLimiter, async (req: Request, res: Response) => {
  try {
    const { publicKey, agentName, signature, challengeId } = req.body;

    if (!publicKey || !signature || !challengeId) {
      res.status(400).json({ error: 'Missing publicKey, signature, or challengeId' });
      return;
    }

    const { account, session } = await accountService.loginAgent(publicKey, agentName, signature, challengeId);

    res.json({
      success: true,
      account: sanitizeAccount(account),
      session: sanitizeSession(session),
    });
  } catch (err: any) {
    res.status(401).json({ error: err.message });
  }
});

/**
 * POST /api/v1/auth/refresh
 * Refresh an expired JWT token.
 */
router.post('/auth/refresh', generalLimiter, async (req: Request, res: Response) => {
  try {
    const { refreshToken } = req.body;

    if (!refreshToken) {
      res.status(400).json({ error: 'Missing refreshToken' });
      return;
    }

    const session = accountService.refreshSession(refreshToken);
    if (!session) {
      res.status(401).json({ error: 'Invalid or expired refresh token' });
      return;
    }

    const account = accountService.getAccount(session.accountId);
    if (!account) {
      res.status(401).json({ error: 'Account not found' });
      return;
    }

    res.json({
      success: true,
      account: sanitizeAccount(account),
      session: sanitizeSession(session),
    });
  } catch (err: any) {
    res.status(401).json({ error: err.message });
  }
});

// ─── Helpers ──────────────────────────────────────────────────

function sanitizeAccount(account: any): any {
  return {
    id: account.id,
    type: account.type,
    email: account.email,
    phone: account.phone,
    displayName: account.displayName,
    agentName: account.agentName,
    publicKey: account.publicKey,
    createdAt: account.createdAt,
    lastLoginAt: account.lastLoginAt,
    status: account.status,
    friendCount: account.friends?.length || 0,
    maxFriends: account.maxFriends,
    friendPermissions: account.friendPermissions || {},
  };
}

function sanitizeSession(session: any): any {
  return {
    sessionId: session.id,
    token: session.token,
    refreshToken: session.refreshToken,
    expiresAt: session.expiresAt,
  };
}

export default router;

File: routes/auth_routes.py
"""
AICQ Auth Routes
==================
Authentication route handlers: verification code generation, registration,
login (human + agent), challenge, and token refresh.

All routes are prefixed with ``/api/v1/auth``.
"""

from __future__ import annotations

import logging
from typing import Any, Dict

from aiohttp import web

from routes.middleware import (
    error_response,
    get_db,
    json_response,
    auth_middleware,
)

logger = logging.getLogger("aicq.auth_routes")


# ─── Route Handlers ─────────────────────────────────────────────────────


async def send_code(request: web.Request) -> web.Response:
    """POST /api/v1/auth/send-code

    Request body::

        {
            "target": "user@example.com",
            "type": "email",          // "email" | "phone"
            "purpose": "register"     // "register" | "login" | "reset_password"
        }

    Generates a 6-digit verification code and sends it to the target.
    """
    db = get_db(request)

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    target = body.get("target")
    vtype = body.get("type")
    purpose = body.get("purpose")

    if not target or not vtype or not purpose:
        return error_response(
            "Missing required fields: target, type, purpose",
            status=400,
            code="MISSING_FIELDS",
        )

    try:
        from services.verification_service import generate_code

        result = await generate_code(db, target, vtype, purpose)
        return json_response(result, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=400, code="VALIDATION_ERROR")
    except Exception as exc:
        logger.exception("Error generating verification code")
        return error_response(
            "Internal server error", status=500, code="INTERNAL_ERROR"
        )


async def register(request: web.Request) -> web.Response:
    """POST /api/v1/auth/register

    Request body::

        {
            "email": "user@example.com",
            "phone": "+1234567890",       // optional
            "password": "secret123",
            "displayName": "Alice",
            "publicKey": "abcd1234...",
            "code": "123456"
        }

    Verifies the email code first, then registers the human account
    and returns account info + tokens.
    """
    db = get_db(request)

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    email = body.get("email")
    phone = body.get("phone")
    password = body.get("password")
    display_name = body.get("displayName", "")
    public_key = body.get("publicKey")
    code = body.get("code")

    if not email or not password or not public_key or not code:
        return error_response(
            "Missing required fields: email, password, publicKey, code",
            status=400,
            code="MISSING_FIELDS",
        )

    # Verify the code first
    try:
        from services.verification_service import verify_code

        verified = await verify_code(db, email, code, "register")
        if not verified:
            return error_response(
                "Invalid or expired verification code",
                status=400,
                code="INVALID_CODE",
            )
    except Exception as exc:
        logger.exception("Error verifying code during registration")
        return error_response(
            "Internal server error", status=500, code="INTERNAL_ERROR"
        )

    # Register the account
    try:
        from services.account_service import register_human, create_session

        account = await register_human(
            db, email, phone, password, display_name, public_key
        )

        # Create session tokens
        tokens = await create_session(db, account["id"])

        return json_response(
            {
                "account": account,
                "access_token": tokens["access_token"],
                "refresh_token": tokens["refresh_token"],
            },
            status=201,
        )
    except ValueError as exc:
        return error_response(str(exc), status=400, code="VALIDATION_ERROR")
    except Exception as exc:
        logger.exception("Error during registration")
        return error_response(
            "Internal server error", status=500, code="INTERNAL_ERROR"
        )


async def login(request: web.Request) -> web.Response:
    """POST /api/v1/auth/login

    Request body::

        {
            "email": "user@example.com",
            "password": "secret123",
            "deviceInfo": "web-chrome"    // optional
        }

    Authenticates a human account and returns tokens.
    """
    db = get_db(request)

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    email = body.get("email")
    password = body.get("password")
    device_info = body.get("deviceInfo")

    if not email or not password:
        return error_response(
            "Missing required fields: email, password",
            status=400,
            code="MISSING_FIELDS",
        )

    try:
        from services.account_service import login_human

        result = await login_human(db, email, password, device_info)
        return json_response(result, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=401, code="INVALID_CREDENTIALS")
    except Exception as exc:
        logger.exception("Error during login")
        return error_response(
            "Internal server error", status=500, code="INTERNAL_ERROR"
        )


async def challenge(request: web.Request) -> web.Response:
    """POST /api/v1/auth/challenge

    Request body::

        {
            "publicKey": "abcd1234..."
        }

    Generates an Ed25519 challenge for agent authentication.
    """
    db = get_db(request)

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    public_key = body.get("publicKey")

    if not public_key:
        return error_response(
            "Missing required field: publicKey",
            status=400,
            code="MISSING_FIELDS",
        )

    try:
        from services.account_service import challenge_agent

        result = await challenge_agent(db, public_key)
        return json_response(result, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=400, code="VALIDATION_ERROR")
    except Exception as exc:
        logger.exception("Error generating challenge")
        return error_response(
            "Internal server error", status=500, code="INTERNAL_ERROR"
        )


async def login_agent(request: web.Request) -> web.Response:
    """POST /api/v1/auth/login-agent

    Request body::

        {
            "publicKey": "abcd1234...",
            "signature": "signed_hex...",
            "challenge": "challenge_hex...",
            "deviceInfo": "agent-v1"      // optional
        }

    Authenticates an AI agent via Ed25519 challenge-response.
    """
    db = get_db(request)

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    public_key = body.get("publicKey")
    signature = body.get("signature")
    challenge = body.get("challenge")
    device_info = body.get("deviceInfo")

    if not public_key or not signature or not challenge:
        return error_response(
            "Missing required fields: publicKey, signature, challenge",
            status=400,
            code="MISSING_FIELDS",
        )

    try:
        from services.account_service import login_agent as _login_agent

        result = await _login_agent(db, public_key, signature, challenge, device_info)
        return json_response(result, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=401, code="INVALID_CREDENTIALS")
    except Exception as exc:
        logger.exception("Error during agent login")
        return error_response(
            "Internal server error", status=500, code="INTERNAL_ERROR"
        )


async def refresh(request: web.Request) -> web.Response:
    """POST /api/v1/auth/refresh

    Request body::

        {
            "refreshToken": "eyJ..."
        }

    Issues a new access/refresh token pair.
    """
    db = get_db(request)

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    refresh_token = body.get("refreshToken")

    if not refresh_token:
        return error_response(
            "Missing required field: refreshToken",
            status=400,
            code="MISSING_FIELDS",
        )

    try:
        from services.account_service import refresh_token as _refresh_token

        result = await _refresh_token(db, refresh_token)
        return json_response(result, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=401, code="INVALID_REFRESH_TOKEN")
    except Exception as exc:
        logger.exception("Error refreshing token")
        return error_response(
            "Internal server error", status=500, code="INTERNAL_ERROR"
        )


# ─── Route Registration ─────────────────────────────────────────────────


def setup_routes(app: web.Application, prefix: str = "/api/v1/auth") -> None:
    """Register all auth routes on the given aiohttp application.

    Parameters
    ----------
    app:
        The aiohttp application instance.
    prefix:
        URL prefix for all auth routes. Defaults to ``/api/v1/auth``.
    """
    app.router.add_post(prefix + "/send-code", send_code)
    app.router.add_post(prefix + "/register", register)
    app.router.add_post(prefix + "/login", login)
    app.router.add_post(prefix + "/challenge", challenge)
    app.router.add_post(prefix + "/login-agent", login_agent)
    app.router.add_post(prefix + "/refresh", refresh)

    logger.info("Auth routes registered under %s", prefix)
